app/assets/javascripts/discourse/app/static/prosemirror/lib/text-manipulation.js

```diff
@@ -0,0 +1,492 @@
+// @ts-check
+import { setOwner } from "@ember/owner";
+import { next } from "@ember/runloop";
+import $ from "jquery";
+import { lift, setBlockType, toggleMark, wrapIn } from "prosemirror-commands";
+import { liftListItem, sinkListItem } from "prosemirror-schema-list";
+import { TextSelection } from "prosemirror-state";
+import { bind } from "discourse/lib/decorators";
+import { i18n } from "discourse-i18n";
+
+/**
+ * @typedef {import("discourse/lib/composer/text-manipulation").TextManipulation} TextManipulation
+ * @typedef {import("discourse/lib/composer/text-manipulation").AutocompleteHandler} AutocompleteHandler
+ * @typedef {import("discourse/lib/composer/text-manipulation").PlaceholderHandler} PlaceholderHandler
+ */
+
+/** @implements {TextManipulation} */
+export default class ProsemirrorTextManipulation {
+  allowPreview = false;
+
+  /** @type {import("prosemirror-model").Schema} */
+  schema;
+  /** @type {import("prosemirror-view").EditorView} */
+  view;
+  $editorElement;
+  /** @type {PlaceholderHandler} */
+  placeholder;
+  /** @type {AutocompleteHandler} */
+  autocompleteHandler;
+  convertFromMarkdown;
+  convertToMarkdown;
+
+  constructor(owner, { schema, view, convertFromMarkdown, convertToMarkdown }) {
+    setOwner(this, owner);
+    this.schema = schema;
+    this.view = view;
+    this.convertFromMarkdown = convertFromMarkdown;
+    this.convertToMarkdown = convertToMarkdown;
+    this.$editorElement = $(view.dom);
```

Comment:
Makes me kinda sad that we have this JQuery object in a class field. I know it's only used for autocomplete... but could we perhaps do the `$(this.view.dom).autocomplete`, so the sadness is contained to a single line, and is more obvious why it's needed?